### ISSUE TITLE
blobs.size() error in FullyConnectedLayerImpl

### ISSUE THREAD

**Issue by TheusStremens** [opencv/opencv#9766]
##### System information (version)
- OpenCV => 3.3
- Operating System / Platform => Ubuntu 16.04
- Compiler => GCC

I'm trying run the LightCNN (https://github.com/AlfredXiangWu/face_verification_experiment more specific the LightenedCNN_C_deploy model) on the new module dnn, the net is loaded normally, but when it does a forward the following error occurs:

"OpenCV Error: Assertion failed (1 <= blobs.size() && blobs.size() <= 2) in FullyConnectedLayerImpl, file /home/matheusm/Git/opencv/modules/dnn/src/layers/fully_connected_layer.cpp, line 61
terminate called after throwing an instance of 'cv::Exception'
  what():  /home/matheusm/Git/opencv/modules/dnn/src/layers/fully_connected_layer.cpp:61: error: (-215) 1 <= blobs.size() && blobs.size() <= 2 in function FullyConnectedLayerImpl"

Here is my code:
```
Mat img = imread(imageFile,0);
// img.convertTo(img, CV_32F, 1.0/255.0);
if (img.empty())
{
    std::cerr << "Can't read image from the file: " << imageFile << std::endl;
    exit(-1);
}
Mat inputBlob = blobFromImage(img, 1, Size(128,128), Scalar(), false);   //Convert Mat to batch of images
// Mat inputBlob = blobFromImage(img);
net.setInput(inputBlob, "data");
Mat descriptor;
descriptor = net.forward("etlwise_fc1");
```

1. probe.jpg is a correct normalized face image with all the requirements from the LightCNN
2. I tried the commented "img.convertTo", the same error. I tried the commented "Mat inputBlob = blobFromImage(img);", the same error.
3. I tried change the layer in the forward, the same error.
4. LightCNN has the same types of layers then GoogleNetCNN, and the sample of classification using GoogleNet works fine
5. LightCNN model and prototxt works fine using Caffe

**Comment by dkurt** [comment]
@TheusStremens, I have a problem with `LightenedCNN_C` model archive so it was tested using generated weights from Caffe. Could you try to run the model with changes from https://github.com/opencv/opencv/pull/9772?

**Comment by TheusStremens** [comment]
I will try and report here

**Comment by TheusStremens** [comment]
I tried your changes @dkurt but the same error appears.

**Comment by dkurt** [comment]
@TheusStremens, please share the .caffemodel because I have no problem with model generated by Caffe.

**Comment by TheusStremens** [comment]
It's the LightenedCNN_C from https://github.com/AlfredXiangWu/face_verification_experiment/tree/master/model

**Comment by dkurt** [comment]
@TheusStremens, the `.prototxt` seems to be broken at the second fully-connected layer: it should has a name `fc2_ms` but not `fc2` so importer can't find associated weights. In Caffe runtime its weights are initialized randomly and you can figure it out by different predictions in different runs:

```
$ python run.py
[[  9.82788408e-20   1.77404289e-18   7.23026172e-14 ...,   2.90189185e-14
    1.47050175e-20   3.32194534e-24]]
$ python run.py
[[  6.30909596e-18   2.24553566e-21   5.22851549e-23 ...,   4.01833357e-16
    5.21777759e-15   2.04026127e-19]]
```
Solution: replace `fc2` to `fc2_ms`
```
$ python run.py
[[  3.63911461e-08   3.62116417e-08   1.71443247e-07 ...,   3.63068047e-08
    3.62632733e-08   3.58028918e-09]]
$ python run.py
[[  3.63911461e-08   3.62116417e-08   1.71443247e-07 ...,   3.63068047e-08
    3.62632733e-08   3.58028918e-09]]
```

```python
# run.py
import caffe
import numpy as np

net = caffe.Net('LightenedCNN_C_deploy.prototxt', 'LightenedCNN_C.caffemodel', caffe.TEST)

np.random.seed(2701)
net.blobs['data'].data[0,0,:,:] = np.random.standard_normal([128, 128]).astype(np.float32)
net.forward()

print net.blobs['prob'].data
```

**Comment by TheusStremens** [comment]
@dkurt replace fc2 to fc2_ms works fine. I don't know why in caffe works without this change and opencv doesn't. About the Caffe weights, they are not initialized randomly, the different predictions in different runs it's because probably you forgot to remove the dropout layer in the LightenedCNN_C_deploy.prototxt. Last thing, of course I cat the tar.gz files, I work with the LightenedCNN_C :smile: . 
Thank you so much for the help and the fix. OpenCV dnn module is very good, it's very fast and straightful, it's amazing to work with CNNs without install a lot of things.

**Comment by dkurt** [comment]
@TheusStremens, you're welcome! Thank you for the feedback! It's important for us.
No, the problem isn't because of Dropout layer: the same thing is observed with it and without it. You may see that Dropout is an identity layer during the `TEST` phase, https://github.com/BVLC/caffe/blob/master/src/caffe/layers/dropout_layer.cpp#L37.
The weights are initialized randomly by fillers:
```
weight_filler {
  type: "xavier"
}
bias_filler {
  type: "constant"
  value: 0.1
}

```
Where `xavier` is a random sampling strategy.

**Comment by dhanyathailappan** [comment]
I too have the same error. Can anyone help me?

OpenCV Error: Assertion failed ((bias && 1->blobs.size() ==2) || (! Bias && 1->blobs.size() ==1) in CV::dnn: :initConvDeconvLayerFromCaffe

And the code link is 
https://www.pyimagesearch.com/2019/02/25/black-and-white-image-colorization-with-opencv-and-deep-learning/

**Comment by myothura2021** [comment]
I have this error. Please help me.

cv2.error: OpenCV(4.5.3) C:\Users\runneradmin\AppData\Local\Temp\pip-req-build-1i5nllza\opencv\modules\dnn\src\layers\fully_connected_layer.cpp:181: error: (-215:Assertion failed) srcMat.dims == 2 && srcMat.cols == weights.cols && dstMat.rows == srcMat.rows && dstMat.cols == weights.rows && srcMat.type() == weights.type() && weights.type() == dstMat.type() && srcMat.type() == CV_32F && (biasMat.empty() || (biasMat.type() == srcMat.type() && biasMat.isContinuous() && (int)biasMat.total() == dstMat.cols)) in function 'cv::dnn::FullyConnectedLayerImpl::FullyConnected::run'

**Comment by phamtho123** [comment]
I have this error. Please help me.
cv2.error: OpenCV(4.3.0) /tmp/pip-install-2g4t1ju8/opencv-python_c4dbb514e5684575b14f0cdc0c2d64a2/opencv/modules/dnn/src/layers/batch_norm_layer.cpp:47: error: (-215:Assertion failed) blobs.size() >= 2 in function 'BatchNormLayerImpl'

**Comment by alalek** [comment]
> 4.3.0

Outdated release (> 1 year).

**Comment by phamtho123** [comment]
requirements only need version 4.2.0
